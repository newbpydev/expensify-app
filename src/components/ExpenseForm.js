import React from "react";

import moment from "moment";
// import { SingleDatePicker } from "react-dates";
// import "react-dates/lib/css/_datepicker.css";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";

// const now = moment();

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      note: props.expense ? props.expense.note : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: "",
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onSubmitHandle = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      // set error message
      this.setState(() => ({
        error: "Please provide description and amount.",
      }));
      if (!this.state.description) {
        // e.target.elements.description.focus();
      } else if (!this.state.amount) {
        e.target.elements.amount.focus();
      }
    } else {
      // clear the error
      this.setState(() => ({
        error: "",
      }));

      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        note: this.state.note,
        createdAt: this.state.createdAt.valueOf(),
      });
    }
  };

  render() {
    return (
      <form className="form" action="" onSubmit={this.onSubmitHandle}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          className="text-input"
          type="text"
          name="description"
          id="description"
          placeholder="Description"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
          // required
        />
        <input
          className="text-input"
          type="text"
          name="amount"
          id="amount"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
          // required
        />

        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          id="tisisis"
          numberOfMonths={1}
          isOutsideRange={() => false}
        />

        <textarea
          className="textarea"
          name="note"
          id="note"
          cols="39"
          rows="5"
          value={this.state.note}
          onChange={this.onNoteChange}
          placeholder="Add a note for your expense (optional)"
        ></textarea>
        <div>
          <button className="button">{this.props.isEditForm ? "Edit" : "Save"} Expense</button>
        </div>
      </form>
    );
  }
}

// const mapStateToProps = (state) => ({
//   expenses: state.expenses
// })

// export default connect(mapStateToProps)(ExpenseForm)
