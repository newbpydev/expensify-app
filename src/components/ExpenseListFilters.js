import React from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByAmount, sortByDate, setEndDate, setStartDate } from "../actions/filters";
import {DateRangePicker} from "react-dates"

class ExpenseListFilters extends React.Component {
  // constructor(props) {
  //   super(props)
  // }
  state = {
    calendarFocused: null,
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };
  onFocusChange = (calendarFocused) =>
    this.setState({ calendarFocused });
  onTextChange = (e) => {
            this.props.dispatch(setTextFilter(e.target.value));
  }
  onSortChange = (e) => {
            if (e.target.value === "date") {
              this.props.dispatch(sortByDate());
            } else if (e.target.value === "amount") {
              this.props.dispatch(sortByAmount());
            }
          }

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              className="text-input"
              type="text"
              name=""
              id=""
              placeholder="Search expenses"
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              name="selectfilter"
              id="select-filter"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
              startDateId="your_unique_start_date_idstart" // PropTypes.string.isRequired,
              endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
              endDateId="your_unique_end_date_idend" // PropTypes.string.isRequired,
              onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
              focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
              onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
              // numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
            />
          </div>
        </div>
      </div>
    );
  }
}

// const ExpenseListFilters = (props) =>

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

export default connect(mapStateToProps)(ExpenseListFilters);
