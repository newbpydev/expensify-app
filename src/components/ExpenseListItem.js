import React from "react";
import moment from "moment";
import numeral from "numeral";

import { Link } from "react-router-dom";

const ExpenseListItem = ({id, description, amount, createdAt, expenses, dispatch }) => (
  <div>
    <h3><Link to={`/edit/${id}`} state={{id}}>{description}</Link></h3>
    <p>
      Amount: {numeral(amount / 100).format("$0,0.00")}
      -
      Created At: {moment(createdAt).format('MMMM Do, YYYY')}
    </p>
    
  </div>
);

// const mapStateToProps = (state) => {
//   return {
//     expenses: state.expenses
//   }
// }

export default ExpenseListItem;
