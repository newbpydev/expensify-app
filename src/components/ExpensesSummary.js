import React from "react";
import { connect } from "react-redux";
import selectExpenses from "../selectors/expenses";
import expensesTotal from "../selectors/expenses-total";
import numeral from "numeral";

export function ExpensesSummary({ dispatch, expensesCount, expensesTotal }) {
  const expenseWord = expensesCount === 1 ? "expense" : "expenses";
  const formattedExpensesTotal = numeral(expensesTotal / 100).format("$0,0.00");
  // console.log(props)
  return (
    <div>
      <h4>
        {expensesCount > 0
          ? `Viewing ${expensesCount} ${expenseWord} totalling ${formattedExpensesTotal}`
          : "There are no expenses"}
      </h4>
    </div>
  );
}

const mapStateToProp = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expensesTotal: expensesTotal(visibleExpenses),
    expensesCount: visibleExpenses.length,
  };
};

export default connect(mapStateToProp)(ExpensesSummary);
