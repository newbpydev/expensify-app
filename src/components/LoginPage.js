import React from "react";
import { useNavigate } from "react-router-dom";

export function LoginPage() {
  return (
    <div>
      <h1>This is the LoginPage</h1>
      <button>Login</button>
    </div>
  )
}







