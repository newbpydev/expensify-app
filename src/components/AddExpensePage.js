import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { startAddExpense, addExpense, editExpense } from "../actions/expenses";

import { useNavigate, Navigate } from "react-router-dom";

// export class AddExpensePage extends React.Component {
//   // navigate = useNavigate(); //! <= IMPORTANT TO ADD BEFORE THE RETURN
//   // STATEMENT
  
//   onSubmit = (expense) => {
//     this.props.onSubmit(expense);
//     // this.navigate("/");
//     // console.log('I am tying to navigate')
//     <Navigate to="/" />;
//   };
  
//   render() {
//     // console.log(props)
//     return (
//       <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm onSubmit={this.onSubmit} />
//       </div>
//     );
//   }
// }
const AddExpensePage = (props) => {
  const navigate = useNavigate(); //! <= IMPORTANT TO ADD BEFORE THE RETURN STATEMENT

  return (
    <div>
      <h1>Add Expense</h1>
      <ExpenseForm
        onSubmit={(expense) => {
          props.dispatch(startAddExpense(expense));
          navigate("/");
        }}
      />
    </div>
  );
};

const mapDispatchToProps = (dispatch) => {
  return {
    startAddExpense: (expense) => dispatch(startAddExpense(expense)),
  };
};

export default connect()(AddExpensePage);
