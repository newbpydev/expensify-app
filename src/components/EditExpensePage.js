import React from "react";
import { useParams } from "react-router-dom";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";
import { useNavigate } from "react-router-dom";

const getExpense = (id, expenses) => {
  return expenses.find((expense) => expense.id === id);
};

const EditExpensePage = (props) => {
  const navigate = useNavigate();
  const { id } = useParams();
  const expenseData = getExpense(id, props.expenses);

  return (
    <div>
      <ExpenseForm
        isEditForm={true}
        expense={expenseData}
        onSubmit={(expense) => {
          props.dispatch(startEditExpense(id, expense));
          navigate("/dashboard");
        }}
      />
      <button
        onClick={(e) => {
          props.dispatch(startRemoveExpense({ id }));
          navigate("/dashboard");
        }}
      >
        Remove
      </button>
    </div>
  );
};

//! Keep the wrapper in mind when using Link with State
// const EditExpensePageWrapper = (props) => {
//   const { id } = useParams()

//   console.log(id)
//   return <EditExpensePage id={id} {...props} />
// }

const mapStateToProps = (state) => {
  return {
    expenses: state.expenses,
  };
};

export default connect(mapStateToProps)(EditExpensePage);
