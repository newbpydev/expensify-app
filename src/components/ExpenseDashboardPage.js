import React from "react";
import { Outlet } from "react-router-dom"
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
