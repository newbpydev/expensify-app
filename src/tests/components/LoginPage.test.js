import React from "react";
import { shallow } from "enzyme";
import {LoginPage} from "../../components/LoginPage";

test("should render the LoginPage correctly", () => {
  const wrapper = shallow(<LoginPage />)

  expect(wrapper).toMatchSnapshot()
})


test("should call startLogin on button action", () => {
  const startLogin = jest.fn(); //! spy to monitor click event
  const wrapper = shallow(<LoginPage startLogin={startLogin} />);

  // expect(wrapper).toMatchSnapshot()
  wrapper.find("#btn-login").simulate("click");
  expect(startLogin).toHaveBeenCalled();
});







