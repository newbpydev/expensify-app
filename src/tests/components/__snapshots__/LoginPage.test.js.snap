// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the LoginPage correctly 1`] = `
<div>
  <h1>
    This is the LoginPage
  </h1>
  <button>
    Login
  </button>
</div>
`;
